torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3693,3 +4203,50 @@ def _is_int(expr):
 # WARNING: This is legacy, DO NOT USE
 def _is_dim_dynamic(t, d):
     return hasattr(t, "_dynamo_dynamic_indices") and d in t._dynamo_dynamic_indices
+
+# Replays the ShapeEnvEvents list.
+# It assumes the first event is the constructor call.
+#
+# fn: transforms an old FX node into one corresponding to the newly created ShapeEnv.
+def replay_shape_env_events(events: List[ShapeEnvEvent]) -> ShapeEnv:
+    constructor_event = events[0]
+    assert constructor_event.f == ShapeEnv
+
+    # Constructs the new ShapeEnv.
+    shape_env = constructor_event.run()
+
+    for event in events[1:]:
+        try:
+            # Actually replays each event.
+            # We need to call create_mapping_fn every time, since the node list might
+            # change after each event is replayed.
+            event.run(shape_env)
+        except Exception as e:
+            raise RuntimeError(f"failed when running event: {event}") from e
+
+    return shape_env
+
+# FakeTensor metadata.
+# This is to be used in place of FakeTensor placeholders when calling
+# ShapeEnv.produce_guards.
+@dataclass
+class FakeTensorMeta:
+    tensor_size: Tuple[Union[int, SymInt], ...]
+    tensor_stride: Tuple[Union[int, SymInt], ...]
+    tensor_storage_offset: Union[int, SymInt]
+
+    def size(self) -> Tuple[Union[int, SymInt], ...]:
+        return self.tensor_size
+
+    def stride(self) -> Tuple[Union[int, SymInt], ...]:
+        return self.tensor_stride
+
+    def storage_offset(self) -> Union[int, SymInt]:
+        return self.tensor_storage_offset
+
+    def dim(self) -> int:
+        return len(self.tensor_size)
+
+    @staticmethod
+    def from_fake(fake) -> "FakeTensorMeta":
+        return FakeTensorMeta(fake.size(), fake.stride(), fake.storage_offset())
```

Comment:
We're cracking 4k lines at this point. Is there any way we can reorganize this into multiple files? 🤔

Reply:
Sure. I will work on it.